### PR TITLE
feat(#8329): API route to measure the effect of purging rules changes

### PR DESCRIPTION
**Pull request by m5r**
# Description

medic/cht-core#8329

- [x] copy purging behavior code from `sentinel` over to `api`
- [ ] write tests to cover the "dry run" code -- in progress
- [ ] merge both `sentinel` and `api` purging code to a shared library -- this includes changes to accept a function to run the queries unique to each project, `sentinel` uses batches of paginated queries because it needs to actually purge documents after querying whereas `api` queries can get away with a single query per type of docs because it's only reading from couchdb views

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:purging-dry-run-api/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:purging-dry-run-api/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:purging-dry-run-api/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: api/src/services/purging-config.js [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+  const willPurgeCount = sumBy(results, 'willPurgeCount');
+  const willUnpurgeCount = sumBy(results, 'willUnpurgeCount');
+
+  const nextRun = getNextRun(appSettingsPurge);
+
+  return { wontChangeCount, willPurgeCount, willUnpurgeCount, nextRun };
+};
+
+module.exports = {
+  dryRun,
```
Comment (kennsippell) [inline api/src/services/purging-config.js:636]: curious about the expected performance for this endpoint? if purging takes several hours in production, how long would this endpoint take to execute? what implications will this have on replication performance?

**Comment by mrjones-plip** [comment]
Since I didn't see it anywhere, adding that Allies is tabling this effort because of the performance concerns both @kennsippell  and @dianabarsan have expressed, as well as the test results @m5r found:

>  I got some disappointing news about our purging dry run solution :disappointed:
> 
> I've started a dry run of a purge in my morning on a clone of Muso-Mali with a beefy machine with similar specs: Xeon E5-2686 v4 @ 2.30GHz, 256 GB of RAM, ~650GB of data stored on a 1.5 TB disk. I'm using a fork of CHT 3.13.0 with the purging dry run API living on the temporary branch 3.13.0-FR-dry-run-purging.
>
>It's the beginning of the night over here and the dry run is still going. It took nearly 5 hours to simulate purging contacts, processing ~10k records with each batched request. Our assumption was that queries were cheap and mutating the data was the expensive part of purging that makes the process so slow but it turns out the queries are expensive as we're seeing roughly the same performances as actual purging despite using couchdb views.
>
>It averages 35% of CPU usage with spikes to 80% and any loss of connection between `cht-conf` and the API during the dry run results in wasted CPU usage as `cht-conf` can't reconnect to the API to wait for the results while the API keeps running the dry run.
>
> With all this, it's safe to say we cannot move forward with this solution and we should go back to the design step for this feature.

**Comment by dianabarsan** [comment]
Thanks a lot for following up on this, @mrjones-plip . Can this PR be closed then?

**Comment by mrjones-plip** [comment]
ah - yeah!  I'll close it for @m5r who is on leave

**Pull request closed without merging**